dc/s2n-quic-dc/src/stream/environment/tokio/pool.rs

```diff
@@ -0,0 +1,234 @@
+}
+
+pub(super) struct Pool {
+    sockets: Box<[Socket]>,
+    current: AtomicUsize,
+    mask: usize,
+}
+
+struct Socket {
+    recv_socket: RecvSocket,
+    application_socket: ApplicationSocket,
+    worker_socket: WorkerSocket,
+    queue: Mutex<Queues>,
+}
+
+impl Socket {
+    fn new(socket: UdpSocket, queue: Queues) -> io::Result<Self> {
+        let recv_socket = Arc::new(socket);
+
+        let local_addr = recv_socket.local_addr()?;
+
+        let send_socket = Tracing(SendOnly(CachedAddr::new(recv_socket.clone(), local_addr)));
+
+        let application_socket = Arc::new(Single(send_socket.clone()));
+
+        let worker_socket = Arc::new(send_socket);
+
+        Ok(Socket {
+            recv_socket,
+            application_socket,
+            worker_socket,
+            queue: Mutex::new(queue),
+        })
+    }
+}
+
+impl Pool {
+    pub fn new<Sub>(env: &Environment<Sub>, workers: usize, mut config: Config) -> Result<Self>
+    where
+        Sub: event::Subscriber,
+    {
+        debug_assert_ne!(workers, 0);
+
+        let workers = workers.next_power_of_two();
+        let mask = workers - 1;
+
+        let mut options = env.socket_options.clone();
+        options.blocking = config.blocking;
+
+        if config.workers.is_none() {
+            config.workers = Some(workers);
+        }
+        let sockets = Self::create_workers(options, &config)?;
+
+        let max_packet_size = config.max_packet_size;
+        let packet_count = config.packet_count;
+        let create_packets = || Packets::new(max_packet_size, packet_count);
+
+        if config.blocking {
+            Self::spawn_blocking(&config.map, &sockets, create_packets)
+        } else {
+            let _rt = env.reader_rt.enter();
+            Self::spawn_non_blocking(&config.map, &sockets, create_packets)?;
+        }
+
+        Ok(Self {
+            sockets: sockets.into(),
+            current: AtomicUsize::new(0),
+            mask,
+        })
+    }
+
+    pub fn alloc(&self) -> (Control, Stream, ApplicationSocket, WorkerSocket) {
+        let idx = self.current.fetch_add(1, Ordering::Relaxed);
+        let idx = idx & self.mask;
+        let socket = &self.sockets[idx];
+        let (control, stream) = socket.queue.lock().unwrap().alloc_or_grow();
+        let app_socket = socket.application_socket.clone();
+        let worker_socket = socket.worker_socket.clone();
+
+        (control, stream, app_socket, worker_socket)
+    }
+
+    fn spawn_blocking(map: &Map, sockets: &[Socket], create_packets: impl Fn() -> Packets) {
+        for (udp_socket_worker, socket) in sockets.iter().enumerate() {
+            let dispatch = socket.queue.lock().unwrap().dispatcher();
+            let dispatch = dispatch.with_map(map.clone());
+            let socket = socket.recv_socket.clone();
+            let packets = create_packets();
+            let span = tracing::trace_span!("udp_socket_worker", udp_socket_worker);
+            std::thread::spawn(move || {
```

Comment:
nit: can we name these threads?

Reply:
yep will do. though with the testing that i've done so far, the blocking impl is slower than the tokio-managed one so it may go away if we can't fix it.